src/edu/csus/ecs/pc2/shadow/RemoteEventFeedMonitor.java

```diff
@@ -251,504 +259,449 @@ public void run() {
-                                                        if (!isReadOnlyClient()) {
-                                                            statusMessage = "Primary started the contest at " + getISODateString(remoteStart);
-                                                            pc2Controller.startAllContestTimes();
-                                                        } else {
-                                                            statusMessage = "Primary indicates contest start at " + getISODateString(remoteStart) +
-                                                                " (R/O client)";
-                                                        }
-                                                    } else {
-                                                        statusMessage = "Primary indicates contest start at " + getISODateString(remoteStart) + " but it's already running";
-                                                    }
-                                                    logAndDebugPrint(log, Level.INFO, statusMessage);
-                                                }
-                                                // If anything related to the startTime decided it wants a status message in the GUI, here is where it comes out
-                                                if(statusMessage != null && monitorStatus != null) {
-                                                    monitorStatus.statusMessage(statusMessage);
+                            try {
+                                logAndDebugPrint(log, Level.INFO, "Found event of type " + notifyType + ": " + dataMap.toString());
+
+                                // Can set this to indicate we want to put a message in the GUI log
+                                String statusMessage = null;
+
+                                //convert state data into a ShadowStateMessage object
+                                ShadowStateMessage contestState = createStateMessage(dataMap);
+
+                                // Keep track of states reported by primary.  We check the various fields in reverse occurrence order.
+                                // As per the spec, the events must appear in order, so we check the last event first, since, if that occurs
+                                // the other must have already happened.
+                                if(contestState.getEndOfUpdates() != null) {
+                                    remoteCCSEOU = generalStateCheck(contestState.getEndOfUpdates(), remoteCCSEOU, "End-of-updates");
+                                } else if(contestState.getFinalized() != null || contestState.getThawed() != null) {
+                                    remoteCCSFinalized = generalStateCheck(contestState.getFinalized(), remoteCCSFinalized, "Finalized");
+                                    remoteCCSThawed = generalStateCheck(contestState.getThawed(), remoteCCSThawed, "Thawed");
```

Comment:
It seems like the various strings being checked for in these statements (e.g. `"End-of-updates"`, "`Finalized`", etc.) should be defined as constants rather than hard-coded here as literal strings.  Otherwise, any changes to their value in the CLICS spec would be harder to maintain.